File: dfs_bfs/boj/p7576.py
#bfs로만 풀어야 하는 문제 (한 방향으로 쭉 익는 것이 아니라 익은 토마토 주변부터 익기 시작해서) 토하고 마시고 토하고..
from sys import stdin
from collections import deque

def bfs(m, n, array, ripe):
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]

    days = -1 #처음 부터 다 익었으면 0 출력해야하므로

    while ripe:
        days += 1 

        for _ in range(len(ripe)):
            x, y = ripe.popleft()

            for j in range(4):
                a = x + dx[j]
                b = y + dy[j]

                if (0 <= a < n) and (0 <= b < m) and (array[a][b] == 0):
                    array[a][b] += 1
                    ripe.append([a, b])
    
    for row in array:
        if 0 in row:
            return -1

    return days

m, n = map(int, stdin.readline().split())
array, ripe = [], deque()

for i in range(n):
    array.append(list(map(int, stdin.readline().split())))

# 이미 익은 토마토들을 queue에 넣어놔야 함
for i in range(n):
    for j in range(m):
        if array[i][j] == 1:
            ripe.append([i,j])

print(bfs(m, n, array, ripe))

File: dfs_bfs/boj/P7576.java
package dfs_bfs.boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

import javax.print.event.PrintEvent;

public class P7576 {
    static int m, n; // m 은 column, n은 row

    static int[][] graph;
    static int[][] visited;

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());

        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        graph = new int[n][m];
        visited = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
                if (graph[i][j] == -1) {
                    visited[i][j] = -1;
                }
            }
        }

        bfs();
        
        int max = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0;  j < m; j++) {
                if (visited[i][j] == 0) {
                    System.out.println("-1");
                    return;
                } else {
                    max = Math.max(max, visited[i][j]);
                }
            }
        }


        // 첫 날을 1로 보니까 -1
        System.out.println(max-1);
    }

    static void bfs() {
        Queue<Node> que = new LinkedList<>();

        // 동시에 연결리스트 찾는 효과!!
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph[i][j] == 1) {
                    que.offer(new Node(i, j));
                    visited[i][j] = 1;
                } 
            }
        }

        while (!que.isEmpty()) {
            Node node = que.poll();

            for (int i = 0; i < 4; i++) {
                int a = node.x + dx[i];
                int b = node.y + dy[i];

                if (0 <= a && a < n && 0 <= b && b < m) {
                    if (graph[a][b] == 0 && visited[a][b] == 0) {
                        que.offer(new Node(a, b));
                        visited[a][b] = visited[node.x][node.y] + 1;
                    }
                }
            }
        }

    }

    static void print() {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                System.out.print(visited[i][j] + " ");
            }
            System.out.println();
        }
    }

    static class Node {
        int x;
        int y;
        
        Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    
}
